### ISSUE TITLE
LiteRT.js / WebGPU: PP-OCRv5 recognizer deterministically misreads text vs WASM (fp32 weights too, Apple Metal)

### ISSUE THREAD

**Issue by john-rocky** [google-ai-edge/LiteRT#9661]
We ship PP-OCRv5 with LiteRT.js in a browser extension (det + rec, [litert-community/PP-OCRv5-LiteRT](https://huggingface.co/litert-community/PP-OCRv5-LiteRT)), and while validating backends we found the recognizer decodes real text crops to different strings on `webgpu` than on `wasm`. The wrong strings are deterministic and unchanged by fp32 weights. The detector from the same release is exact and fast on WebGPU, so we ship det on WebGPU and rec on WASM to work around it.

### Environment

- **LiteRT.js version**: 2.5.3 (threaded WASM variant, `numThreads: 8`)
- **Browser**: Chromium 153.0.8006.0 (first observed on Chrome 151 stable, same 2.5.3)
- **OS / GPU**: macOS 27.0, Apple M4 Max — WebGPU adapter `vendor=apple`, `architecture=metal-3`
- **Model**: `ppocr_rec_fp16.tflite` and `ppocr_rec_fp32.tflite` (its pre-quantization parent) from [litert-community/PP-OCRv5-LiteRT](https://huggingface.co/litert-community/PP-OCRv5-LiteRT), input `[1,3,48,320]` float32

### Steps to reproduce

```bash
git clone https://github.com/john-rocky/page-voice
cd page-voice && npm install
node repro/serve.mjs   # open http://127.0.0.1:8123/ocr-rec-webgpu-misdecode/ in Chrome
```

The page loads both models on both backends, feeds identical tensors built from committed crop PNGs (self-authored text, pixel-identical everywhere), CTC-decodes, and prints the comparison ([repro source](https://github.com/john-rocky/page-voice/tree/main/repro/ocr-rec-webgpu-misdecode)).

### Actual result

On 5 of 35 rendered test crops (EN/JA, several font sizes), the decoded string differs between backends. Examples:

| crop | wasm | webgpu |
|---|---|---|
| ![crop](https://raw.githubusercontent.com/john-rocky/page-voice/main/repro/ocr-rec-webgpu-misdecode/crops/crop-02-en-28px.png) | `Hello OCR 2026` | `YHello 0CR 2026` |
| ![crop](https://raw.githubusercontent.com/john-rocky/page-voice/main/repro/ocr-rec-webgpu-misdecode/crops/crop-04-ja-28px.png) | `日本語対応です` | `a本語対応です` |

Observations from the same runs, all on this machine:

- **Deterministic**: two WebGPU runs per crop are id-for-id identical, on all 35 crops (fp16 model).
- **Not fp16 rounding**: the fp32-weight model produces the *same wrong strings* on WebGPU (4 of the 5 crops, byte-identical to the fp16 output). `gpuOptions: {precision: 'fp32'}` on the fp16 model changed nothing either (measured 2026-08-14).
- **Errors cluster at the line start**: a phantom first character (`Y`, or 日→`a`), then mostly correct text. Inserting background columns before the content does not move or remove them (measured 2026-08-14).
- **Control**: the PP-OCRv5 detector from the same conversion ([1,3,640,640], random input, same comparison method, measured 2026-08-11) matches WebGPU-vs-WASM to ~3e-11 max-abs — the divergence is specific to the recognizer graph. Two control crops decode identically on all four model×backend combinations.
- On random input, the recognizer's raw logits differ by up to 0.084 max-abs between backends (measured 2026-08-11, same versions).

Extent: one machine (Apple Metal); not yet tried on other GPUs. Full end-to-end (det→rec) comparisons show the same pattern on 6 of 24 pipeline lines.

### Expected result

Same decoded string on both backends (up to numeric noise below CTC decision level, as the detector and the two control crops show).

Possibly related: #8065 reports wrong WebGPU output on Adreno, but with a different signature (correct first run, garbage afterwards, non-deterministic); here the wrong output is deterministic from the first run. #9662 is a numeric divergence on another model's graph on the same machine.

Thanks for looking — happy to run further probes or candidate fixes on this machine; the repro takes about a minute end to end.

**Comment by shivvaam0001** [comment]
Hi @john-rocky,

Thank you for reaching out to us. We will investigate this further and get back to you.

**Comment by shivvaam0001** [comment]
Hi @john-rocky,

I tested your repro locally on Apple Silicon (M3 / Metal 3, Chromium 152, @litertjs/core@2.5.3) and saw the same behavior:

- 5 out of 7 crops produce different results on WebGPU, but the results are consistent across repeated runs (`webgpuDeterministic: true`).
- Same issue with FP32: I also tested `ppocr_rec_fp32.tflite` and got the same corrupted tokens ("YHello 0CR 2026", "a本語対応です"), so this does not seem to be related to FP16 quantization.
- The detector results are almost identical between WebGPU and WASM (max difference ~3e-11), and the control crops match 100%.

**Comment by aravindmurali22** [comment]
Hi @pyu10055  , following up on this thread, is there any recent updates on this? Thanks!
